src/envoy/http/authn/http_filter.cc

```diff
@@ -48,49 +49,36 @@ FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap& headers,
   filter_context_.reset(new Istio::AuthN::FilterContext(
       &headers, decoder_callbacks_->connection(), filter_config_));
 
-  authenticator_ = createPeerAuthenticator(
-      filter_context_.get(),
-      [this](bool success) { onPeerAuthenticationDone(success); });
-  authenticator_->run();
+  Payload payload;
 
-  if (state_ == State::COMPLETE) {
-    return FilterHeadersStatus::Continue;
+  if (!createPeerAuthenticator(filter_context_.get())->run(&payload)) {
```

Comment:
This is just a small step of simplifying thing.  But we still have so many classes:  Context, base, Peer, and Origin. 

 Can we further simplify it into one class:

// with all data you need
Authn  authn(policy,  header, filter_callback);

if (authn.Check()) {
    authn.get_payload_result();
} else {
   authn.get_error_message.
}

Reply:
That I want to keep. Our auth policy do have the concept of peer and origin, so it's easier to see implementation follow that hierarchy.

Context also make implementation independent of filter, which make testing much easier and cleaner.